File: c-code/global_dot.c
#include <stdio.h>
#include <stdlib.h>
#include <mpi.h>

int main(int argc, char** argv) {
    int rank, size, i, n = 100;
    int *x = NULL, *y = NULL, *subx = NULL, *suby = NULL;
    int local_dot = 0, global_dot = 0;

    MPI_Init(&argc, &argv); // Initialize MPI environment
    MPI_Comm_rank(MPI_COMM_WORLD, &rank); // Get the rank of the current process
    MPI_Comm_size(MPI_COMM_WORLD, &size); // Get the total number of processes

    // Allocate memory for the arrays
    if (rank == 0) {
        x = (int*) malloc(n * sizeof(int));
        y = (int*) malloc(n * sizeof(int));
        for (i = 0; i < n; i++) {
            x[i] = i + 1;
            y[i] = n - i;
        }
    }

    // Scatter the arrays to all processes
    subx = (int*) malloc(n / size * sizeof(int));
    suby = (int*) malloc(n / size * sizeof(int));
    MPI_Scatter(x, n / size, MPI_INT, subx, n / size, MPI_INT, 0, MPI_COMM_WORLD);
    MPI_Scatter(y, n / size, MPI_INT, suby, n / size, MPI_INT, 0, MPI_COMM_WORLD);

    // Compute the local dot product of the subarrays
    for (i = 0; i < n / size; i++) {
        local_dot += subx[i] * suby[i];
    }

    // Reduce the local dot products to obtain the global dot product
    MPI_Reduce(&local_dot, &global_dot, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

    // Print the result on the root process
    if (rank == 0) {
        printf("The dot product of the arrays is %d\n", global_dot);
        free(x);
        free(y);
    }
    free(subx);
    free(suby);

    MPI_Finalize(); // Finalize MPI environment
    return 0;
}

File: python-code/global_dot.py
from mpi4py import MPI
import numpy as np

comm = MPI.COMM_WORLD
rank = comm.rank
size = comm.size
n = 2 * size

if rank == 0:
    x = np.arange(1, n+1)
    y = np.arange(1, n+1)
    chunks_x = [[] for i in range(size)]
    chunks_y = [[] for i in range(size)]
    # data needs to be divided in chunks manually for scatter
    for i in range(size):
        for j in range(n//size):
            chunks_x[i].append(x[i*(n//size)+j])
            chunks_y[i].append(y[i*(n//size)+j])
else:
    x = None
    y = None
    chunks_x = None
    chunks_y = None

# Scatter the data array to all processes
chunks_x = comm.scatter(chunks_x, root=0)
chunks_y = comm.scatter(chunks_y, root=0)
print("Process", rank, "has received ", chunks_x, "and", chunks_y)

# Compute the local sum of the subdata array
local_dot = sum([chunks_x[i] * chunks_y[i] for i in range(len(chunks_x))])

# Reduce the local sum to obtain the global sum
global_dot = comm.reduce(local_dot, op=MPI.SUM, root=0)

# Print the result on the root process
if rank == 0:
    print("The sum of the array is", global_dot)
